airflow/contrib/hooks/sagemaker_hook.py

```diff
@@ -16,299 +16,793 @@
-            else:
-                running = False
+    def get_log_conn(self):
+        """
+        Establish an AWS connection for retrieving logs during training
 
-        self.log.info('SageMaker Job Compeleted')
+        :return: a boto3 CloudWatchLog client
+        """
+        config = botocore.config.Config(retries={'max_attempts': 15})
+        return self.get_client_type('logs', config=config)
 
-    def get_conn(self):
+    def get_iam_conn(self):
         """
-        Establish an AWS connection
-        :return: a boto3 SageMaker client
+        Establish an AWS connection for retrieving IAM roles during training
+
+        :return: a boto3 IAM client
         """
-        return self.get_client_type('sagemaker', region_name=self.region_name)
+        return self.get_client_type('iam')
 
-    def list_training_job(self, name_contains=None, status_equals=None):
+    def log_stream(self, log_group, stream_name, start_time=0, skip=0):
         """
-        List the training jobs associated with the given input
-        :param name_contains: A string in the training job name
-        :type name_contains: str
-        :param status_equals: 'InProgress'|'Completed'
-        |'Failed'|'Stopping'|'Stopped'
-        :return:dict
+        A generator for log items in a single stream. This will yield all the
+        items that are available at the current moment.
+
+        :param log_group: The name of the log group.
+        :type log_group: str
+        :param stream_name: The name of the specific stream.
+        :type stream_name: str
+        :param start_time: The time stamp value to start reading the logs from (default: 0).
+        :type start_time: int
+        :param skip: The number of log entries to skip at the start (default: 0).
+            This is for when there are multiple entries at the same timestamp.
+        :type skip: int
+        :return:A CloudWatch log event with the following key-value pairs:
+            'timestamp' (int): The time of the event.
+            'message' (str): The log event data.
+            'ingestionTime' (int): The time the event was ingested.
         """
-        return self.conn.list_training_jobs(
-            NameContains=name_contains, StatusEquals=status_equals)
 
-    def list_tuning_job(self, name_contains=None, status_equals=None):
+        next_token = None
+
+        event_count = 1
+        while event_count > 0:
+            if next_token is not None:
+                token_arg = {'nextToken': next_token}
+            else:
+                token_arg = {}
+
+            response = self.get_log_conn().get_log_events(logGroupName=log_group,
+                                                          logStreamName=stream_name,
+                                                          startTime=start_time,
+                                                          startFromHead=True,
+                                                          **token_arg)
```

Comment:
Yep I was surprised when I did not find it during development. My concern for 'filter_log_events' is, it does not have an argument to force events order by timestamp. All docs I can find did not mention it. 

I do need log events to be in time order since I am extracting these logs during runtime. SageMaker keeps writing logs to cloudwatch and my codes keep extracting them to Airflow Log. I have to make sure what users see it in timestamp order.

Let me confirm with cloudwatch team for this. Since at least my approach now is not wrong (it's basically a customized paginator), I probably can update this in next PR after I get response fro cloudwatch guys?

Reply:
Sounds good

Reply:
I talked with cloudwatch guys. They identified my use case as 'tailing the logs'. So 'get_log_events' is what they recommend to use (nextToken will never be null in this case unlike what in 'filter_log_events').

Reply:
Hence I guess I don't need to change these codes although I don't like them :) compared to a paginator.